# Segmenting and Clustering Neighborhoods in Toronto

# Introduction

### Let's explore neighborhoods in Toronto, Canada using Foursquare API. We will explore the most common venues in Neighborhoods and group them into clusters

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [78]:
# Importing Libraries
import requests
!conda install -c anaconda beautifulsoup4 --yes
from bs4 import BeautifulSoup
from urllib.request import urlopen
import pandas as pd

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - beautifulsoup4


The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    conda-forge::ca-certificates-2020.6.2~ --> anaconda::ca-certificates-2020.1.1-0
  certifi            conda-forge::certifi-2020.6.20-py36h9~ --> anaconda::certifi-2020.6.20-py36_0
  openssl            conda-forge::openssl-1.1.1g-h516909a_0 --> anaconda::openssl-1.1.1g-h7b6447c_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done


## 1. Download and Explore Dataset

For the Toronto neighborhood data, a Wikipedia page exists, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M that has all the information we need to explore and cluster the neighborhoods in Toronto. We will scrape the Wikipedia page into a structured format using Pandas Dataframe


The data is available in table format with Postal code, Borough and Neighborhood as columns

In [79]:
# Using pandas to read the table into pandas dataframe
df=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', flavor='html5lib')
# Extract the first table in the page
df_neighborhood=df[0]  
df_neighborhood.head() 

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

The table consists of few cells with no Borough. Remove the cells in which Borough is Not assigned

In [80]:
df_neighborhood.drop(df_neighborhood[df_neighborhood['Borough']=='Not assigned'].index, inplace=True)
df_neighborhood.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

chech the shape of the dataframe

In [81]:
df_neighborhood.shape

(103, 3)

The dataframe has 103 postal codes. The table is grouped by postal codes. There are multiple Borough and Neighborhoods with same postal code. 
So we will go ahead and segment using postal codes.

Check for any duplicate postal codes or empty Borough cells

In [82]:
bool_series =df_neighborhood["Postal Code"].duplicated().any()
print(bool_series)

False


## 2. Download the Spatial coordinates for all postal codes

The spatial coordinates for all postal codes are available in 'Geospatial_Coordinates.csv'. Using the spatial coordinates we will explore the venues for all neighborhoods.


Load the Spatial coordinates in to pandas dataframe

In [83]:
df_coordinates=pd.read_csv('Geospatial_Coordinates.csv')
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Merge the 2 dataframes grouped by Postal codes

In [84]:
df_toronto = pd.merge(df_neighborhood, df_coordinates, on='Postal Code', how='outer')
df_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [85]:
df_toronto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postal Code   103 non-null    object 
 1   Borough       103 non-null    object 
 2   Neighborhood  103 non-null    object 
 3   Latitude      103 non-null    float64
 4   Longitude     103 non-null    float64
dtypes: float64(2), object(3)
memory usage: 4.8+ KB


## 3. Foursquare API

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [86]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20200101' # Foursquare API version

print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:


#### Now, let's get the top 100 venues that are in Parkwoods within a radius of 500 meters.

In [87]:

LIMIT = 100 # limit of number of venues returned by Foursquare API


radius = 500 # define radius

In [88]:
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    43.753259, 
    -79.329656, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=KQYJCCZ32VQRTL1EJIASG4YTWEFP4R2TWNBIRFW2ITX2UMQA&client_secret=DPQOCMUFXJ1HXLECKTNMWJVBXSFSG0RNFW31SYUNRFQXMYME&v=20200101&ll=43.753259,-79.329656&radius=500&limit=100'

In [89]:
results = requests.get(url).json()["response"]['groups'][0]['items']
results

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
   'name': 'Brookbanks Park',
   'location': {'address': 'Toronto',
    'lat': 43.751976046055574,
    'lng': -79.33214044722958,
    'labeledLatLngs': [{'label': 'display',
      'lat': 43.751976046055574,
      'lng': -79.33214044722958}],
    'distance': 245,
    'cc': 'CA',
    'city': 'Toronto',
    'state': 'ON',
    'country': 'Canada',
    'formattedAddress': ['Toronto', 'Toronto ON', 'Canada']},
   'categories': [{'id': '4bf58dd8d48988d163941735',
     'name': 'Park',
     'pluralName': 'Parks',
     'shortName': 'Park',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'groups': []}},
  'referralId': 'e-0-4e8d9dcdd5fbbbb6b3003c7b-0'},
 {'reasons': {'count': 0,
   'items': [{'

We can see that all the information is in the *items* key.Lets extract the venues for all neifghborhoods

In [90]:
def getNearbyVenues(postal_code,neighborhood, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, neigh, lat, lng in zip(postal_code,neighborhood, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        try:
            results = requests.get(url).json()["response"]["groups"][0]['items']
        except:
            KeyError: 'groups'
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            neigh,
            lat, 
            lng, 
            v['venue']['name'], 
            #v['venue']['location']['lat'], 
            #v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code',
                             'Neighborhood',
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *Toronto_venues*.

In [91]:
Toronto_venues = getNearbyVenues(postal_code=df_toronto['Postal Code'],
                                 neighborhood=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

In [92]:
Toronto_venues.head()

Postal Code      Neighborhood  Neighborhood Latitude  \
0         M3A         Parkwoods              43.753259   
1         M3A         Parkwoods              43.753259   
2         M3A         Parkwoods              43.753259   
3         M4A  Victoria Village              43.725882   
4         M4A  Victoria Village              43.725882   

   Neighborhood Longitude                              Venue  \
0              -79.329656                    Brookbanks Park   
1              -79.329656                      Variety Store   
2              -79.329656  Corrosion Service Company Limited   
3              -79.315572             Victoria Village Arena   
4              -79.315572                          Portugril   

               Venue Category  
0                        Park  
1           Food & Drink Shop  
2  Construction & Landscaping  
3                Hockey Arena  
4       Portuguese Restaurant

Lets check the shape of the Toronto_venues

In [93]:
Toronto_venues.shape

(2130, 6)

Lets check the total number of unique venue categories

In [94]:

print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 272 uniques categories.


In [95]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Let's check how many venues were returned for each neighborhood

In [96]:
Toronto_groups=Toronto_venues.groupby('Neighborhood').count()

In [97]:
Toronto_groups

Postal Code  \
Neighborhood                                                      
Agincourt                                                     5   
Alderwood, Long Branch                                        8   
Bathurst Manor, Wilson Heights, Downsview North              20   
Bayview Village                                               4   
Bedford Park, Lawrence Manor East                            23   
Berczy Park                                                  56   
Birch Cliff, Cliffside West                                   5   
Brockton, Parkdale Village, Exhibition Place                 24   
Business reply mail Processing Centre, South Ce...           16   
CN Tower, King and Spadina, Railway Lands, Harb...           16   
Caledonia-Fairbanks                                           4   
Canada Post Gateway Processing Centre                        13   
Cedarbrae                                                     8   
Central Bay Street                                           64   
Christie                                                     17   
Church and Wellesley                                         75   
Clarks Corners, Tam O'Shanter, Sullivan                      15   
Cliffside, Cliffcrest, Scarborough Village West               2   
Commerce Court, Victoria Hotel                              100   
Davisville                                                   33   
Davisville North                                              9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn              4   
Don Mills                                                    26   
Dorset Park, Wexford Heights, Scarborough Town ...            7   
Downsview                                                    17   
Dufferin, Dovercourt Village                                 16   
East Toronto, Broadview North (Old East York)                 3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...            9   
Fairview, Henry Farm, Oriole                                 66   
First Canadian Place, Underground city                      100   
Forest Hill North & West, Forest Hill Road Park               6   
Garden District, Ryerson                                    100   
Glencairn                                                     5   
Golden Mile, Clairlea, Oakridge                               9   
Guildwood, Morningside, West Hill                             7   
Harbourfront East, Union Station, Toronto Islands           100   
High Park, The Junction South                                25   
Hillcrest Village                                             4   
Humber Summit                                                 2   
Humberlea, Emery                                              1   
Humewood-Cedarvale                                            4   
India Bazaar, The Beaches West                               24   
Kennedy Park, Ionview, East Birchmount Park                   5   
Kensington Market, Chinatown, Grange Park                    62   
Kingsview Village, St. Phillips, Martin Grove G...            4   
Lawrence Manor, Lawrence Heights                             15   
Lawrence Park                                                 4   
Leaside                                                      33   
Little Portugal, Trinity                                     43   
Malvern, Rouge                                                1   
Milliken, Agincourt North, Steeles East, L'Amor...            2   
Mimico NW, The Queensway West, South of Bloor, ...           14   
Moore Park, Summerhill East                                   3   
New Toronto, Mimico South, Humber Bay Shores                 12   
North Park, Maple Leaf Park, Upwood Park                      4   
North Toronto West, Lawrence Park                            20   
Northwest, West Humber - Clairville                           3   
Northwood Park, York University                               5   
Old Mill South, King's Mill Park

In [98]:
Toronto_groups.shape

(95, 5)

## 3. Analyze Each Neighborhood

In [102]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe

Toronto_onehot.drop('Neighborhood', axis=1, inplace=True)
Toronto_onehot.insert(0,'Neighborhood',Toronto_venues['Neighborhood'])


Toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2         Parkwoods                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bike Shop  Bistro  \
0                   0                         0          0       0   
1                   0                         0          0       0   
2                   0                         0          0       0   
3                   0                         0          0       0   
4                   0                         0          0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bub

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [103]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.043478   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0               0.0                 0.0          0.0            0.0   
1               0.0                 0.0          0.0            0.0   
2               0.0                 0.0          0.0            0.0   
3               0.0                 0.0          0.0            0.0   
4               0.0                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0        0.0         0.0         0.0     0.0  0.00  0.0             0.0   
1        0.0         0.0         0.0     0.0  0.00  0.0             0.0   
2        0.0         0.0         0.0     0.0  0.10  0.0             0.0   
3        0.0         0.0         0.0     0.0  0.25  0.0             0.0   
4        0.0         0.0         0.0     0.0  0.00  0.0             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0               0.0               0.0                 0.0    0.0   
1               0.0               0.0                 0.0    0.0   
2               0.0               0.0                 0.0    0.0   
3               0.0               0.0                 0.0    0.0   
4               0.0               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0              0.0       0.0         0.0                 0.0   
1              0.0       0.0         0.0                 0.0   
2              0.0       0.0         0.0                 0.0   
3              0.0       0.0         0.0                 0.0   
4              0.0       0.0         0.0                 0.0   

   Bike Rental / Bike Share  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                       0.0        0.0     0.0         0.0            0.0   
1                       0.0        0.0     0.0         0.0            0.0   
2                       0.0        0.0     0.0         0.0            0.0   
3                       0.0        0.0     0.0         0.0            0.0   
4                       0.0        0.0     0.0         0.0            0.0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0        0.0       0.0                   0.0             0.2      0.0   
1

#### Let's confirm the new size

In [104]:
Toronto_grouped.shape

(95, 272)

First, let's write a function to sort the venues in descending order.

In [105]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [106]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Clothing Store                Lounge        Breakfast Spot   
1                   Gym          Skating Rink                  Pool   
2           Coffee Shop           Pizza Place           Supermarket   
3                  Café                  Bank   Japanese Restaurant   
4           Coffee Shop    Italian Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0          Skating Rink   Dumpling Restaurant  Eastern European Restaurant   
1                   Pub        Sandwich Place                  Coffee Shop   
2      Sushi Restaurant         Deli / Bodega                   Restaurant   
3               Dog Run    Dim Sum Restaurant                        Diner   
4             Juice Bar       Thai Restaurant                 Liquor Store   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Drugstore            Donut Shop          Deli / Bodega  
1                    Gay Bar      Doner Restaurant    Distribution Center  
2  Middle Eastern Restaurant     Mobile Phone Shop            Bridal Shop  
3             Discount Store   Distribution Center            Yoga Studio  
4          Indian Restaurant                   Pub                Butcher

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [107]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [118]:
# add clustering labels
neighborhoods_venues_sorted.drop('Cluster Labels', axis=1, inplace=True)
neighborhoods_venues_sorted.insert(1, 'Cluster Labels', kmeans.labels_)
Toronto_merged=neighborhoods_venues_sorted

Toronto_merged = Toronto_merged.join(df_toronto.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() 

Neighborhood  Cluster Labels  \
0                                        Agincourt               0   
1                           Alderwood, Long Branch               0   
2  Bathurst Manor, Wilson Heights, Downsview North               0   
3                                  Bayview Village               0   
4                Bedford Park, Lawrence Manor East               0   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant        Clothing Store                Lounge   
1                Pizza Place                   Gym          Skating Rink   
2                       Bank           Coffee Shop           Pizza Place   
3         Chinese Restaurant                  Café                  Bank   
4             Sandwich Place           Coffee Shop    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Breakfast Spot          Skating Rink   Dumpling Restaurant   
1                  Pool                   Pub        Sandwich Place   
2           Supermarket      Sushi Restaurant         Deli / Bodega   
3   Japanese Restaurant               Dog Run    Dim Sum Restaurant   
4            Restaurant             Juice Bar       Thai Restaurant   

         7th Most Common Venue      8th Most Common Venue  \
0  Eastern European Restaurant                  Drugstore   
1                  Coffee Shop                    Gay Bar   
2                   Restaurant  Middle Eastern Restaurant   
3                        Diner             Discount Store   
4                 Liquor Store          Indian Restaurant   

  9th Most Common Venue 10th Most Common Venue Postal Code      Borough  \
0            Donut Shop          Deli / Bodega         M1S  Scarborough   
1      Doner Restaurant    Distribution Center         M8W    Etobicoke   
2     Mobile Phone Shop            Bridal Shop         M3H   North York   
3   Distribution Center            Yoga Studio         M2K   North York   
4                   Pub                Butcher         M5M   North York   

    Latitude  Longitude  
0  43.794200 -79.262029  
1  43.602414 -79.543484  
2  43.754328 -79.442259  
3  43.786947 -79.385975  
4  43.733283 -79.419750

## 5. Examine Clusters

#### Cluster 1

In [126]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0]

Neighborhood  Cluster Labels  \
0                                           Agincourt               0   
1                              Alderwood, Long Branch               0   
2     Bathurst Manor, Wilson Heights, Downsview North               0   
3                                     Bayview Village               0   
4                   Bedford Park, Lawrence Manor East               0   
5                                         Berczy Park               0   
6                         Birch Cliff, Cliffside West               0   
7        Brockton, Parkdale Village, Exhibition Place               0   
8   Business reply mail Processing Centre, South C...               0   
9   CN Tower, King and Spadina, Railway Lands, Har...               0   
11              Canada Post Gateway Processing Centre               0   
12                                          Cedarbrae               0   
13                                 Central Bay Street               0   
14                                           Christie               0   
15                               Church and Wellesley               0   
16            Clarks Corners, Tam O'Shanter, Sullivan               0   
17    Cliffside, Cliffcrest, Scarborough Village West               0   
18                     Commerce Court, Victoria Hotel               0   
19                                         Davisville               0   
20                                   Davisville North               0   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn               0   
22                                          Don Mills               0   
22                                          Don Mills               0   
23  Dorset Park, Wexford Heights, Scarborough Town...               0   
24                                          Downsview               0   
24                                          Downsview               0   
24                                          Downsview               0   
24                                          Downsview               0   
25                       Dufferin, Dovercourt Village               0   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...               0   
28                       Fairview, Henry Farm, Oriole               0   
29             First Canadian Place, Underground city               0   
30    Forest Hill North & West, Forest Hill Road Park               0   
31                           Garden District, Ryerson               0   
32                                          Glencairn               0   
33                    Golden Mile, Clairlea, Oakridge               0   
34                  Guildwood, Morningside, West Hill               0   
35  Harbourfront East, Union Station, Toronto Islands               0   
36                      High Park, The Junction South               0   
37                                  Hillcrest Village               0   
38                                      Humber Summit               0   
40                                 Humewood-Cedarvale               0   
41                     India Bazaar, The Beaches West               0   
42        Kennedy Park, Ionview, East Birchmount Park               0   
43          Kensington Market, Chinatown, Grange Park               0   
44  Kingsview Village, St. Phillips, Martin Grove ...               0   
45                   Lawrence Manor, Lawrence Heights               0   
46                                      Lawrence Park               0   
47                                            Leaside               0   
48                           Little Portugal, Trinity               0   
51  Mimico NW, The Queensway West, South of Bloor,...               0   
53       New Toronto, Mimico South, Humber Bay Shores               0   
54           North Park, Maple Leaf Park, Upwood Park               0   
55                  North Toronto West, Lawrence Park               0   
56                Northwes

#### Cluster 2

In [127]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1]

Neighborhood  Cluster Labels 1st Most Common Venue  \
94  York Mills, Silver Hills               1             Cafeteria   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
94           Yoga Studio   Distribution Center      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
94          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
94        Discount Store               Dog Run           Dance Studio   

   Postal Code     Borough  Latitude  Longitude  
94         M2L  North York  43.75749 -79.374714

#### Cluster 3

In [128]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2]

Neighborhood  Cluster Labels  \
39                                   Humberlea, Emery               2   
58  Old Mill South, King's Mill Park, Sunnylea, Hu...               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
39        Baseball Field           Yoga Studio      Department Store   
58        Baseball Field           Yoga Studio      Department Store   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
39          Dessert Shop    Dim Sum Restaurant                 Diner   
58          Dessert Shop    Dim Sum Restaurant                 Diner   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
39        Discount Store   Distribution Center               Dog Run   
58        Discount Store   Distribution Center               Dog Run   

   10th Most Common Venue Postal Code     Borough   Latitude  Longitude  
39            Event Space         M9M  North York  43.724766 -79.532242  
58            Event Space         M8Y   Etobicoke  43.636258 -79.498509

#### Cluster 4

In [129]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3]

Neighborhood  Cluster Labels  \
10                                Caledonia-Fairbanks               3   
26      East Toronto, Broadview North (Old East York)               3   
50  Milliken, Agincourt North, Steeles East, L'Amo...               3   
52                        Moore Park, Summerhill East               3   
65                                           Rosedale               3   
87                                             Weston               3   
93                                    York Mills West               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                  Park                Bakery         Women's Store   
26                  Park     Convenience Store   Distribution Center   
50                  Park            Playground   Distribution Center   
52                  Park            Playground            Restaurant   
65                  Park                 Trail            Playground   
87                  Park   Distribution Center         Deli / Bodega   
93                  Park     Convenience Store   Distribution Center   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
10               Airport         Deli / Bodega  Ethiopian Restaurant   
26      Department Store          Dessert Shop    Dim Sum Restaurant   
50         Deli / Bodega      Department Store          Dessert Shop   
52                 Diner          Dance Studio         Deli / Bodega   
65         Deli / Bodega      Department Store          Dessert Shop   
87      Department Store          Dessert Shop    Dim Sum Restaurant   
93      Department Store          Dessert Shop    Dim Sum Restaurant   

   7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
10     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
26                 Diner               Discount Store               Dog Run   
50    Dim Sum Restaurant                        Diner        Discount Store   
52      Department Store                 Dessert Shop    Dim Sum Restaurant   
65    Dim Sum Restaurant                        Diner        Discount Store   
87                 Diner               Discount Store               Dog Run   
93                 Diner               Discount Store               Dog Run   

   10th Most Common Venue Postal Code           Borough   Latitude  Longitude  
10              Drugstore         M6E              York  43.689026 -79.453512  
26           Dance Studio         M4J         East York  43.685347 -79.338106  
50                Dog Run         M1V       Scarborough  43.815252 -79.284577  
52    Distribution Center         M4T   Central Toronto  43.689574 -79.383160  
65    Distribution Center         M4W  Downtown Toronto  43.679563 -79.377529  
87          Grocery Store         M9N              York  43.706876 -79.518188  
93           Dance Studio         M2P        North York  43.752758 -79.400049

#### Cluster 5

In [130]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4]

Neighborhood  Cluster Labels 1st Most Common Venue  \
49  Malvern, Rouge               4  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
49           Yoga Studio               Dog Run      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
49          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
49        Discount Store   Distribution Center       Doner Restaurant   

   Postal Code      Borough   Latitude  Longitude  
49         M1B  Scarborough  43.806686 -79.194353

In [111]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim


Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be UPDATED:

  ca-certificates      anaconda::ca-certificates-2020.1.1-0 --> conda-forge::ca-certificates-2020.6.20-hecda079_0

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi                anaconda::certifi-2020.6.20-py36_0 --> conda-forge::certifi-2020.6.20-py36h9f0ad1d_0
  openssl               anaconda::openssl-1.1.1g-h7b6447c_0 --> conda-forge::openssl-1.1.1g-h516909a_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done


In [112]:
import matplotlib.cm as cm
import matplotlib.colors as colors
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [113]:
import folium
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 6. Conclusion

1) Toronto is a city with many parks , Gym, Yoga and outdoor activities meaning it is a calm,
relaxed city and people also focused on their fitness and health.

2) There are not many Neighborhoods with cluster of Restaurants or venues suggesting not a lot of activities or tourism in Toronto

3) Toronto seems to be diverse since there are many neighborhoods with Asian and Latin American Restaurants.